master/buildbot/process/buildrequestmerger.py

```diff
@@ -0,0 +1,246 @@
+        """
+        # Never merge if a build request has a selected_slave
+        # This might happen when a user wants to test the same build in different
+        # slaves to look for instabilities
+        if 'selected_slave' in properties:
+            return None
+
+        # Get an initial list of all breqs of the same name, in the same chain
+        matchingBreqs = self._getSameBuildersInChain(startbrid, builderName)
+
+        # Get properties for matching breqs (done in a single queyr for optimization)
+        otherProperties = self._getBuildsetsProperties(
+            [bsid for bsid, _brid in matchingBreqs])
+
+        # Check if relevant properties match
+        for otherBuildsetid, otherBrid in matchingBreqs:
+            if self._propertiesMatch(properties,
+                                     otherProperties[otherBuildsetid],
+                                     mergeProperties):
+
+                # If they match, merge against this buildrequest
+                return otherBrid
+
+        # If we can't find any match, return None
+        return None
+
+    def _getSameBuildersInChain(self, startbrid, builderName):
+        """
+        :param str startbrid:
+        :param str builderName:
+        :return tuple(str,str):
+            (buildsetid, brid) for all buildrequests in the same chain (`startbrid`)
+            with the same `builderName`
+        """
+        breq_tbl = self.master.db.model.buildrequests
+
+        def __getSameBuildersInChain(conn):
+            # Select buildrequests `buildsetit` and `brid`
+            # q = sa.select(breq_tbl.c.buildsetid, breq_tbl.c.id) \
+
+            # For builders in the same chain
+            #     .where(breq_tbl.c.startbrid == startbrid) \
+
+            # With the same name
+            #     .where(breq_tbl.c.buildername == builderName) \
+
+            # That were not merged themselves (only merge against one / main target)
+            #     .where(breq_tbl.c.mergebrid == None) \
+            q = sa.select(breq_tbl.c.buildsetid, breq_tbl.c.id) \
+                .where(breq_tbl.c.startbrid == startbrid) \
+                .where(breq_tbl.c.buildername == builderName) \
+                .where(breq_tbl.c.mergebrid == None)
+            res = conn.execute(q)
+            return res.fetchall() or []
+
+        return self.master.db.pool.do(__getSameBuildersInChain)
+
+    def _propertiesMatch(self, properties, otherProperties, mergeProperties):
+        """
+        :param dict(str,str) properties:
+        :param dict(str,str) otherProperties:
+        :param list(str) mergeProperties:
+            List of properties that must match
+        :return bool:
+            True if `properties` and `otherProperties` match for a list of
+            `mergeProperties` to be checked
+        """
+        if 'selected_slave' in otherProperties:
+            return False
+
+        for propertyName in self.BASE_MERGE_PROPERTIES + mergeProperties:
```

Comment:
it doesn't look like we handle the case where mergeProperties is None

Reply:
It should never be `None`, it defaults to `[]`